feat: attachment carousel for generic media files

sanjeevdialpad opened the pull request:
# Attachment carousel

<!--- Feel free to remove any unused sections -->

## :hammer_and_wrench: Type Of Change

<!--- Tick or place an `x` in all of the checkboxes that apply -->

- [ ] Fix
- [x] Feature
- [ ] Refactoring
- [ ] Documentation

## :book: Description

Created the attachment carousel recipe that will be used in message input. This can handle any type of media files. The list of media files include: images, video, files.

## :bulb: Context

This PR contains changes pertaining to attachment carousel. For now i have used the generic DtImageViewer but it needs to be replaced with generic type that can handle all different types of media.
Each of these media type will also have a loader and preload state defined.

Figma: https://www.figma.com/file/GJiRmWORmDi0zxLJfwn8yp/DP-2.0---Conversation-View?type=design&node-id=2649-89072&mode=design&t=xHQr2H4s3pCjWKeU-0
JIRA: https://dialpad.atlassian.net/browse/DLT-1159

I will be creating separate PR to be merged into this PR for each component type.

## :pencil: Checklist

<!--- Tick or place an `x` in all of the checkboxes that apply -->
<!--- Remove checkboxes that do not apply -->

- [ ] I have reviewed my changes
- [ ] I have added tests
- [ ] I have added all relevant documentation
- [ ] I have validated components with a screen reader
- [ ] I have validated components keyboard navigation
- [ ] I have considered the performance impact of my change
- [ ] I have checked that my change did not significantly increase bundle size
- [ ] I am exporting any new components or constants in the index.js in the component directory
- [ ] I am exporting any new components or constants in the index.js in the root

## :crystal_ball: Next Steps

<!--- Describe any future changes that need to be made after merging the PR -->

## :camera: Screenshots / GIFs

<!--- Mandatory for any UI work -->
<!--- Link any screenshots / GIFs below -->

## :link: Sources

<!--- Add any links to external reference material -->

sanjeevdialpad committed:
feat: attachment carousel with images

sanjeevdialpad committed:
fix

sanjeevdialpad committed:
revert image viewer changes and make arrow unfocusable

braddialpad commented:
> The list of media files include: images, video, files.

What happens in the case of a user uploading a text file or an audio file or something? It doesn't use this carousel?

sanjeevdialpad commented:
@braddialpad I just checked dialpad upload, it only allows video/Image file. not even Text files. Do we want to support all types of files for this recipe?

I will ask our PM on what the requirement would be for dp 2.0

braddialpad commented:
> @braddialpad I just checked dialpad upload, it only allows video/Image file. not even Text files. Do we want to support all types of files for this recipe?

Fair enough, I think this is not great functionality, but if our goal is just to reach current DP 1.0 functionality then this is fine. Definitely don't let it hold back this PR. possible future update though.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.mdx:
This purposely not done yet?

sanjeevdialpad replied:
ye am probably gonna finish the documentation, variants and test cases in a separate PR. For now i will redo this with the most basic default story in it and description

braddialpad replied:
k sounds good.

sanjeevdialpad replied:
added the docs for base style.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+        @mouseenter="closeButton(true, index)"
+        @mouseleave="closeButton(false, index)"
+      >
+        <dt-image-viewer
+          image-button-class="d-h64 d-w64 d-bar4 d-ba d-baw6 d-bc-subtle"
+          :image-src="mediaItem.path"
+          :image-alt="mediaItem.altText"
+          :close-aria-label="closeAriaLabel"
+          :aria-label="clickToOpenAriaLabel"
+        />
```
Maybe more of an image viewer problem, but should the image maintain its aspect ratio instead of being cut off like this?

![Screenshot 2023-09-26 at 11 34 35 AM](https://github.com/dialpad/dialtone-vue/assets/64808812/a2049def-d912-4af6-8109-b555e688445b)

sanjeevdialpad replied:
yea i noticed this. we manually set the image class but i think image viewer should have a way to set the height and width mainting the aspect ratio somehow passing that value as prop. if you think that idea is fine, we can create a ticket and fix this later.

iropolo replied:
maybe we can use the `object-fit: cover` css property

braddialpad replied:
> yea i noticed this. we manually set the image class but i think image viewer should have a way to set the height and width mainting the aspect ratio somehow passing that value as prop. if you think that idea is fine, we can create a ticket and fix this later.

I don't think it needs to be that complicated, it should probably just always maintain the aspect ratio.

sanjeevdialpad replied:
> maybe we can use the `object-fit: cover` css property

this works perfectly.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+        />
+      </template>
+    </dt-button>
+    <dt-button
+      v-if="showRightArrow && showCarouselArrows"
+      tabindex="-1"
+      class="dt-attachment-carousel--right-arrow d-ba d-baw1 d-bc-default"
+      circle
+      size="xs"
+      importance="clear"
+      :aria-label="rightArrowAriaLabel"
+      @click="rightScroll"
+    >
```
Does the arrow need to dynamically reposition when you delete images?

![Screenshot 2023-09-26 at 11 36 59 AM](https://github.com/dialpad/dialtone-vue/assets/64808812/914a3c6e-5b40-4de4-a606-42c4d66bda68)

sanjeevdialpad replied:
damn, i actually have this condition on remove media to handle this bug but seems like theres more corner cases. Let me try out.
```
removeMediaItem (index) {
      // make sure the carousel arrows is updated. 64 is the width of each media item
      this.showRightArrow = this.$refs.carousel.scrollWidth > (this.$refs.carousel.clientWidth + 64);
      this.$emit('remove-media', index);
    },
    ```

sanjeevdialpad replied:
oh damn, i put that code in the other PR. Let me move it here

braddialpad replied:
Looks good after update.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+import { DtIcon } from '@/components/icon';
+import { DtImageViewer } from '@/components/image_viewer';
+import { DtButton } from '@/components/button';
+import {} from './attachment_carousel_constants';
```
can remove this if not used

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+      @scroll="handleScroll"
+    >
+      <!-- Image list -->
+      <div
+        v-for="(mediaItem, index) in mediaList"
+        :key="`media-${index}`"
+        role="presentation"
+        class="d-ps-relative"
+        @focusin="closeButton(true, index)"
+        @focusout="closeButton(false, index)"
+        @mouseenter="closeButton(true, index)"
+        @mouseleave="closeButton(false, index)"
+      >
```
Hmm something seems messed with keyboard functionality. Tabbing between them makes the focus go to some "hidden" place before going to the next item. Maybe it is going to the close button but it's invisible if it is.

sanjeevdialpad replied:
Yea this is something i wasnt able to fix. Still trying. It doesnt tab to the close button. Coz i kept a focus listener and it didnt get triggered. But i do see this issue.

sanjeevdialpad replied:
solved using opacity to show hide buttons

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+    </div>
+
+    <!-- Carousel Arrows -->
+    <dt-button
+      v-if="showLeftArrow && showCarouselArrows"
+      tabindex="-1"
+      class="dt-attachment-carousel--left-arrow d-ba d-baw1 d-bc-default"
+      circle
+      size="xs"
+      importance="clear"
+      :aria-label="leftArrowAriaLabel"
+      @click="leftScroll"
+    >
+      <template #icon>
+        <dt-icon
+          name="arrow-left"
+          size="100"
+        />
+      </template>
+    </dt-button>
+    <dt-button
+      v-if="showRightArrow && showCarouselArrows"
+      tabindex="-1"
+      class="dt-attachment-carousel--right-arrow d-ba d-baw1 d-bc-default"
+      circle
+      size="xs"
+      importance="clear"
+      :aria-label="rightArrowAriaLabel"
+      @click="rightScroll"
+    >
```
don't think these arrows need an aria label right? They are for visual purposes only.

sanjeevdialpad replied:
right. removed.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+    @focusin="showCarouselArrows = true"
+    @focusout="showCarouselArrows = false"
+  >
+    <div
+      v-if="mediaList.length > 0"
+      ref="carousel"
+      class="dt-attachment-carousel--media-list"
+      :aria-label="attachmentAriaLabel"
+      @scroll="handleScroll"
+    >
```
accessbiility warning here

```
.dt-attachment-carousel--media-list

Serious
aria-label attribute is not well supported on a div with no valid role attribute.
```

is there a better semantic element or role?

sanjeevdialpad replied:
i just put this on the parent one since it has a role presentation. but theres accessbility issue on progress bar am gonna fix.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+          :image-src="mediaItem.path"
+          :image-alt="mediaItem.altText"
+          :close-aria-label="closeAriaLabel"
+          :aria-label="clickToOpenAriaLabel"
```
Should you add the alt text in here so it says "click to open" and then a description of the image?

sanjeevdialpad replied:
`image-alt`  is passed onto the image viewer alt text. Is that what you are talking about?

braddialpad replied:
I was thinking something like.. ``:image-alt="`Click to open ${mediaItem.altText}`"``

But I realized this is an i18n issue so nevermind.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+        @mouseleave="closeButton(false, index)"
+      >
+        <dt-image-viewer
+          image-button-class="d-h64 d-w64 d-bar4 d-ba d-baw6 d-bc-subtle"
```
make a custom class for this using tokens, we try not to use utility classes in dialtone vue anymore.

sanjeevdialpad replied:
changed it in both PRs.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+    <dt-button
+      v-if="showLeftArrow && showCarouselArrows"
+      tabindex="-1"
+      class="dt-attachment-carousel--left-arrow d-ba d-baw1 d-bc-default"
```
Remove utility classes and add them into the custom class (same goes for any others in this PR)

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+  },
+
+  mounted: function () {
+    this.showLeftArrow = this.$refs.carousel.scrollLeft > 0;
+    this.showRightArrow = this.$refs.carousel.scrollWidth > this.$refs.carousel.clientWidth;
```
can you just call handleScroll?

sanjeevdialpad replied:
the showRightArrow is different in handleScroll. Once mounted, we check if there is more items than what can fit in the width. While on scroll we check if we reached the last item.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.vue:
```diff
@@ -0,0 +1,263 @@
+  background-color: var(--dt-color-black-400);
+}
+.dt-attachment-carousel--left-arrow {
+  z-index: 100;
```
Do you need z-index here at all or does it work without?

sanjeevdialpad replied:
works without it. removing.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel.stories.js:
```diff
@@ -0,0 +1,118 @@
+  args: {},
+};
+
+const VariantsTemplate = (args) => createTemplateFromVueFile(
+  args,
+  DtRecipeAttachmentCarouselVariantsTemplate,
+);
+
+export const Variants = {
+  render: VariantsTemplate,
+  args: {},
+};
```
Can remove variants if it's not any different than default.

braddialpad commented on recipes/conversation_view/attachment_carousel/attachment_carousel_variants.story.vue:
Remove file if not needed.

braddialpad reviewed (changes requested):
Lookin pretty good!

sanjeevdialpad committed:
fix: review comments

sanjeevdialpad committed:
fix: review comments

sanjeevdialpad committed:
fix: border color tokens

sanjeevdialpad committed:
feat(attachment-carousel): Image thumbnail with progress bar for the attachment carousel. (#1212)

sanjeevdialpad committed:
fix: a11 test fix

sanjeevdialpad committed:
fix: a11 test fix

sanjeevdialpad committed:
fix: tab to close button fix

sanjeevdialpad committed:
fix: unique close button id

sanjeevdialpad committed:
Merge branch 'staging-vue3' into DLT-1159

braddialpad committed:
fix(root-layout): convert to pure CSS grid method (#1254)

* fix(root-layout): convert to pure CSS grid method

* separate classes

* fix storybook booleans

* fix test

* map fixed prop correctly

* remove sidebar sticky

semantic-release-bot committed:
chore(release): 3.89.0

# [3.89.0](https://github.com/dialpad/dialtone-vue/compare/v3.88.1...v3.89.0) (2023-10-13)

### Bug Fixes

* **Root Layout:** convert to pure CSS grid method ([#1254](https://github.com/dialpad/dialtone-vue/issues/1254)) ([3dbd0be](https://github.com/dialpad/dialtone-vue/commit/3dbd0be47fdf4717fa226bfda1485a8ca54f4fe9))

### Features

* **Rich Text Editor:** contact mentions ([#1248](https://github.com/dialpad/dialtone-vue/issues/1248)) ([19a99fb](https://github.com/dialpad/dialtone-vue/commit/19a99fb2f8de554b83997b232b1e0a2c0f726e24))

sanjeevdialpad committed:
fix: make it highly accessible

sanjeevdialpad committed:
Merge remote-tracking branch 'origin/staging-vue3' into DLT-1159

sanjeevdialpad commented:
@braddialpad this is ready for final review!

braddialpad reviewed (approved):
Looks good once you fix the accessibility error!

sanjeevdialpad committed:
fix: accessibility arrow

sanjeevdialpad committed:
Merge branch 'staging-vue3' into DLT-1159

sanjeevdialpad committed:
Merge branch 'staging-vue3' into DLT-1159
